pallets/liquidation/src/lib.rs

```diff
@@ -0,0 +1,295 @@
+		///
+		/// Performs a flash loan to get funds to pay for the debt.
+		/// Received collateral is swapped and the profit is transferred to `FeeReceiver`.
+		///
+		/// Parameters:
+		/// - `origin`: Signed origin.
+		/// - `collateral_asset`: Asset ID used as collateral in the MM position.
+		/// - `debt_asset`: Asset ID used as debt in the MM position.
+		/// - `user`: EVM address of the MM position that we want to liquidate.
+		/// - `debt_to_cover`: Amount of debt we want to liquidate.
+		/// - `route`: The route we trade against. Required for the fee calculation.
+		///
+		/// Emits `Liquidated` event when successful.
+		///
+		#[pallet::call_index(0)]
+		#[pallet::weight(<T as Config>::WeightInfo::liquidate()
+			.saturating_add(<T as Config>::RouterWeightInfo::sell_weight(route))
+			.saturating_add(<T as Config>::GasWeightMapping::gas_to_weight(<T as Config>::GasLimit::get(), true))
+		)]
+		pub fn liquidate(
+			origin: OriginFor<T>,
+			collateral_asset: AssetId,
+			debt_asset: AssetId,
+			user: EvmAddress,
+			debt_to_cover: Balance,
+			route: Vec<Trade<AssetId>>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let pallet = Self::account_id();
+
+			let debt_original_balance = <T as Config>::Currency::balance(debt_asset, &pallet);
+			let collateral_original_balance = <T as Config>::Currency::balance(collateral_asset, &pallet);
+
+			// mint debt asset
+			<T as Config>::Currency::mint_into(debt_asset, &pallet, debt_to_cover)?;
+
+			// liquidation call
+			let pallet_address = T::EvmAccounts::evm_address(&pallet);
+			let contract = BorrowingContract::<T>::get();
+
+			let context = CallContext::new_call(contract, pallet_address);
+			let data = Self::encode_liquidation_call_data(collateral_asset, debt_asset, user, debt_to_cover, false);
+
+			let (exit_reason, value) = T::Evm::call(context, data, U256::zero(), T::GasLimit::get());
+			if exit_reason != ExitReason::Succeed(ExitSucceed::Returned) {
+				log::debug!(target: "liquidation",
+					"Evm execution failed. Reason: {:?}", value);
+				return Err(Error::<T>::LiquidationCallFailed.into());
+			}
+
+			// swap collateral if necessary
+			if collateral_asset != debt_asset {
+				let collateral_earned = <T as Config>::Currency::balance(collateral_asset, &pallet)
+					.checked_sub(collateral_original_balance)
+					.ok_or(ArithmeticError::Underflow)?;
+				T::Router::sell(
+					RawOrigin::Signed(pallet.clone()).into(),
+					collateral_asset,
+					debt_asset,
+					collateral_earned,
+					1,
+					route,
+				)?;
+			}
+
+			// burn debt and transfer profit
+			let debt_gained = <T as Config>::Currency::balance(debt_asset, &pallet)
+				.checked_sub(debt_original_balance)
+				.ok_or(Error::<T>::NotProfitable)?;
+
+			let profit = debt_gained
+				.checked_sub(debt_to_cover)
+				.ok_or(Error::<T>::NotProfitable)?;
+
+			<T as Config>::Currency::burn_from(debt_asset, &pallet, debt_to_cover, Precision::Exact, Fortitude::Force)?;
```

Comment:
i havent checked integration tests - is there a test that verifies total issuance of the asset bfore and after the liquidate call ?

Reply:
Yes, this check is in the tests.

Reply:
but not in integration tests, but in unit tests